src/wayland/input_method/input_method.cpp

```diff
@@ -0,0 +1,124 @@
+#include "input_method.hpp"
+#include <cstdint>
+
+#include <qdebug.h>
+#include <qlogging.h>
+#include <qobject.h>
+#include <qpointer.h>
+#include <qstring.h>
+#include <qtmetamacros.h>
+#include <qwayland-input-method-unstable-v2.h>
+#include <wayland-input-method-unstable-v2-client-protocol.h>
+
+#include "keyboard_grab.hpp"
+#include "manager.hpp"
+#include "types.hpp"
+
+namespace qs::wayland::input_method::impl {
+
+InputMethodHandle::InputMethodHandle(QObject* parent, ::zwp_input_method_v2* inputMethod)
+    : QObject(parent)
+    , zwp_input_method_v2(inputMethod) {}
+
+InputMethodHandle::~InputMethodHandle() {
+	this->sendPreeditString("");
+	this->destroy();
+}
+
+void InputMethodHandle::commitString(const QString& text) { this->commit_string(text); }
+void InputMethodHandle::sendPreeditString(
+    const QString& text,
+    int32_t cursorBegin,
+    int32_t cursorEnd
+) {
+	this->set_preedit_string(text, cursorBegin, cursorEnd);
+}
+void InputMethodHandle::deleteText(int before, int after) {
+	this->zwp_input_method_v2::delete_surrounding_text(before, after);
+}
+void InputMethodHandle::commit() { this->zwp_input_method_v2::commit(this->serial++); }
+
+bool InputMethodHandle::hasKeyboard() const { return this->keyboard != nullptr; }
+QPointer<InputMethodKeyboardGrab> InputMethodHandle::grabKeyboard() {
```

Comment:
There is no point in returning a QPointer. QPointers are registered to the QObject on creation to receive a hook on its destruction and null themselves. As the hook can't do anything else they're generally not very useful, but they are never useful at all in this position.